include/deal.II/grid/grid_tools.h

```diff
@@ -2142,6 +2162,453 @@ namespace GridTools
+                    break;
+                  }
+                else
+                  {
+                    guess_weights[vertex_n] = 1.0/distance;
+                    guess_weights_sum += guess_weights[vertex_n];
+                  }
+              }
+            guess_weights /= guess_weights_sum;
+            Assert(weights_are_ok(guess_weights), ExcInternalError());
+
+            // Gradient descent:
+            for (unsigned int outer_n = 0; outer_n < 10; ++outer_n)
+              {
+                const unsigned int dependent_direction = GeometryInfo<structdim>::vertices_per_cell - 1;
+                Tensor<1, GeometryInfo<structdim>::vertices_per_cell> current_gradient;
+                for (unsigned int row_n = 0;
+                     row_n < GeometryInfo<structdim>::vertices_per_cell;
+                     ++row_n)
+                  {
+                    if (row_n != dependent_direction)
+                      {
+                        current_gradient[row_n] = gradient_entry<spacedim, structdim>
+                                                  (row_n,
+                                                   dependent_direction,
+                                                   trial_point,
+                                                   guess_weights,
+                                                   step_size,
+                                                   get_point_from_weights);
+
+                        current_gradient[dependent_direction] -= current_gradient[row_n];
+                      }
+                  }
+
+                // since we need to travel in the -gradient direction, guess that
+                // we will go -0.5*gradient and do quasi-Newton iteration to pick
+                // the best multiplier:
+                double gradient_weight = -0.5;
+                auto gradient_weight_objective_function = [&](const double gradient_weight_guess)
+                                                          -> double
+                {
+                  return (trial_point -
+                  get_point_from_weights(guess_weights +
+                  gradient_weight_guess*current_gradient)).norm_square();
+                };
+
+                for (unsigned int inner_n = 0; inner_n < 10; ++inner_n)
+                  {
+                    const double update_numerator = centered_first_difference
+                                                    (gradient_weight, step_size, gradient_weight_objective_function);
+                    const double update_denominator = centered_second_difference
+                                                      (gradient_weight, step_size, gradient_weight_objective_function);
+
+                    // avoid division by zero. Note that we limit the gradient weight below
+                    if (std::abs(update_denominator) == 0.0)
+                      break;
+                    gradient_weight -= update_numerator/update_denominator;
```

Comment:
I don't think I understand this line. Isn't it the case that you're doing a kind of line search here? If so, where is `inner_n` used to determine a step length? And why do we *update* instead of *set* this quantity within the iteration?

Reply:
Maybe it would be useful to outline the algorithm you have here in formulas in a comment above